test/Microsoft.AspNetCore.Blazor.Build.Test/RenderingRazorIntegrationTest.cs

```diff
@@ -598,5 +598,70 @@ public void SupportsTwoWayBindingForEnumValues()
         }
 
         public enum MyEnum { FirstValue, SecondValue }
+
+        [Fact]
+        public void RazorTemplate_CanBeUsedFromRazorCode()
+        {
+            // Arrange
+            var component = CompileToComponent(@"
+@{ RenderFragment<string> template = @<div>@item.ToLower()</div>; }
+@for (var i = 0; i < 3; i++)
+{
+    @template.Bind(""Hello, World!"");
+}
+");
+
+            // Act
+            var frames = GetRenderTree(component);
+
+            // Assert
+            Assert.Collection(
+                frames,
+                frame => AssertFrame.Element(frame, "div", 2, 0),
+                frame => AssertFrame.Text(frame, "hello, world!", 1),
+                frame => AssertFrame.Element(frame, "div", 2, 0),
+                frame => AssertFrame.Text(frame, "hello, world!", 1),
+                frame => AssertFrame.Element(frame, "div", 2, 0),
+                frame => AssertFrame.Text(frame, "hello, world!", 1));
+        }
+
+        [Fact]
+        public void RazorTemplate_CanBeUsedFromMethod()
+        {
+            // Arrange
+            var component = CompileToComponent(@"
+@(Repeat(@<div>@item.ToLower()</div>, ""Hello, World!"", 3))
+
+@functions {
+    RenderFragment Repeat<T>(RenderFragment<T> template, T value, int count)
+    {
+        return (b) =>
+        {
+            for (var i = 0; i < count; i++)
+            {
+                template(b, value);
+            }
+        };
+    }
+}");
+
+            // Act
+            var frames = GetRenderTree(component);
+
+            // Assert
+            //
+            // Review: @stevesa
```

Comment:
@SteveSandersonMS

Reply:
This is totally fine and matches how the sequence numbers look in a regular `@for` or `@foreach` loop. The diff algorithm notices when the sequence number steps back to an earlier number, and uses that to detect the case where leading or trailing items are added within a given iteration.